Upstream arm64e ptrauth changes from Apple's libffi

jeremyhu opened the pull request:
This reverts a recent incomplete change for arm64e support (#548) and includes all potentially upstreamable changes to libffi made in macOS 10.15.5 and iOS 13.4, including Apple's arm64e support.

Note: **Please do not merge this yet**.  It is here for comments first as I also want to hear from @oleavr and others from #548 .

I have not tested these changes against the current master, I have only tested them with our libffi-26 tag based on c2a6859012d928b67a83619bd5087674a96b9254

Note that I have omitted our Xcode project changes, but if there is interest in upstreaming those as well, please let me know.

oleavr commented on src/aarch64/internal.h:
```diff
@@ -65,3 +65,24 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+/* Helpers for writing assembly compatible with arm ptr auth */
+#ifdef LIBFFI_ASM
+
+#if __has_feature(ptrauth_calls)
```
(See earlier note about GCC support.)

oleavr commented on src/aarch64/sysv.S:
```diff
@@ -86,9 +78,22 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
+
 	/* Use a stack frame allocated by our caller.  */
-	cfi_def_cfa(x1, 32);
+#if __has_feature(ptrauth_calls) && defined(__APPLE__)
```
(See earlier note about GCC support.)

oleavr commented on include/ffi.h.in:
```diff
@@ -357,6 +357,11 @@ ffi_prep_closure_loc (ffi_closure*,
 		      void *user_data,
 		      void*codeloc);
 
+#if defined(__x86_64__) || defined(__arm64__)
```
Would it make sense to make this available on all architectures?

jeremyhu replied:
If this seems valuable outside of our internal use case, then I certainly can.  One of the main reasons I'm putting all this up is to see if any of this work is valuable to the larger OSS community or should remain Apple-internal changes for our specific cases.

We only wrote the implementation in src/aarch64/ffi.c and arc/x86/ffi64.c since that is all we use, but it should be usable elsewhere.  This was necessitated by the trampoline changes, and I think we ended up using it in python3 and a few other places where there were assumptions that codeloc == closure.

oleavr replied:
Ahh I see. That's a good question for the maintainers. /cc @atgreen @tromey

atgreen replied:
Leave it for now.  We may come back to this as we need to do something for the Linux architectures.

jeremyhu replied:
Ok.  I'll break that out into a separate PR that adds the API to all archs.

@atgreen What are your opinions on the API naming?  Is ffi_find_closure_for_code() ok, or would you prefer something different?

oleavr commented on src/aarch64/trampoline.S:
```diff
@@ -0,0 +1,44 @@
+    .endr
+
+#endif /* FFI_EXEC_TRAMPOLINE_TABLE */
+#endif /* __arm64 */
```
Build system changes seem to be missing for this file.

jeremyhu replied:
Yeah, the build system changes were specific to our Xcode project.  I'm not looking forward to figuring that out for autotools.

oleavr replied:
Ah no worries, I can take a stab at it and post a suggestion here.

oleavr replied:
Just pushed a rough draft [here](https://github.com/frida/libffi/tree/apple-arm64e).

Here's how I tested it:

### Generate a build environment

```sh
$ git clone https://github.com/frida/frida.git
$ cd frida
$ make build/frida-env-ios-arm64e.rc
```

> Piggybacking on the autotools legacy build environment in [Frida](https://frida.re/), which we still use for building a few dependencies that haven't yet moved to [Meson](https://mesonbuild.com/).

### Bootstrap and build for iOS/arm64e

```
$ . build/frida-env-ios-arm64e.rc
$ cd ~/src/libffi
$ ./autogen.sh
$ ./configure --prefix=/usr --enable-shared
$ make
$ make install DESTDIR="$(echo ~/Desktop/libffi-ios-arm64e)"
```

> --enable-shared is only needed because Frida's $CONFIG_SITE defaults to enable_shared=no

### Build for macOS

From a freshly initialized shell:

```sh
$ . ~/src/frida/build/frida-env-macos-x86_64.rc
$ ./configure --prefix=/usr --enable-shared
$ make
$ make install DESTDIR="$(echo ~/Desktop/libffi-macos-x86_64)"
```

jeremyhu replied:
Cool, thanks for the help.  FWIW, I am *crazy* busy with WWDC work and taking care of kids at home, but I plan to circle back to this after WWDC, break out some of the smaller less offensive pieces into individual PRs, and then polish up the rest.

oleavr commented on src/closures.c:
```diff
@@ -198,34 +220,57 @@ ffi_trampoline_table_alloc (void)
+  static void *ffi_closure_trampoline_table_page;
+
+  dispatch_once(&trampoline_template_init_once, ^{
+    void * const trampoline_handle = dlopen("/usr/lib/libffi-trampolines.dylib", RTLD_NOW | RTLD_LOCAL | RTLD_FIRST);
```
This should take `libdir` into account instead of hardcoding it.

jeremyhu replied:
Yep, good point.

oleavr commented on src/closures.c:
```diff
@@ -198,34 +220,57 @@ ffi_trampoline_table_alloc (void)
+    void * const trampoline_handle = dlopen("/usr/lib/libffi-trampolines.dylib", RTLD_NOW | RTLD_LOCAL | RTLD_FIRST);
+    assert(trampoline_handle);
+
+    ffi_closure_trampoline_table_page = dlsym(trampoline_handle, "ffi_closure_trampoline_table_page");
```
```suggestion
    ffi_closure_trampoline_table_page = dlsym (trampoline_handle, "ffi_closure_trampoline_table_page");
```

Here and elsewhere. It looks like the coding style has eroded somewhat since the early days, but this seems to be the most prevalent style.

oleavr commented on src/closures.c:
```diff
@@ -146,15 +146,22 @@ ffi_closure_free (void *ptr)
-extern void *ffi_closure_trampoline_table_page;
+#if __has_feature(ptrauth_calls)
+#include <ptrauth.h>
+#define sign_ptr(p) ptrauth_sign_unauthenticated(p, ptrauth_key_function_pointer, 0)
```
Unused macro?

jeremyhu commented on src/aarch64/ffi.c:
```diff
@@ -31,6 +31,10 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 #include <windows.h> /* FlushInstructionCache */
 #endif
 
+#if __has_feature(ptrauth_calls)
```
Yeah, I mentioned that already.  I just wanted to get this up for discussion.  The final version will definitely include the existing configure.ac-foo and HAVE_PTRAUTH.

atgreen commented on src/prep_cif.c:
```diff
@@ -234,7 +234,7 @@ ffi_status ffi_prep_cif_var(ffi_cif *cif,
 }
 
-#if FFI_CLOSURES
+#if FFI_CLOSURES && FFI_LEGACY_CLOSURE_API
```
What is FFI_LEGACY_CLOSURE_API?

jeremyhu replied:
In our build, it's defined in our ffitarget.hs as:

darwin/include/ffitarget_arm64.h
54:#define FFI_LEGACY_CLOSURE_API 0

darwin/include/ffitarget_armv7.h
68:#define FFI_LEGACY_CLOSURE_API 0

darwin/include/ffitarget_x86.h
125:#define FFI_LEGACY_CLOSURE_API 1

We just didn't want to make ffi_prep_closure() available on arm since it won't work there... better to catch issues at build time than at runtime.  We should omit these bits from the final PR.

atgreen commented on src/aarch64/ffi.c:
```diff
@@ -31,6 +31,10 @@ SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 #include <windows.h> /* FlushInstructionCache */
 #endif
 
+#if __has_feature(ptrauth_calls)
```
Thanks. I just noticed this caused a  build failure here: https://travis-ci.org/github/libffi/libffi/jobs/692670895#L652

jeremyhu replied:
Cool!  I love CI that works =)  I've updated the title to make it more clear that I'm soliciting feedback on what you think is interesting to upstream rather than merge this in wholesale.  Parts that you find interesting, I'll definitely work to get integrated with autotools builds and cross-platform (as best I can at least) and spin off separate PRs for isolated changes.

atgreen replied:
I put a lot of effort into getting the CI to work for lots of platforms, even implementing a new tool (https://rl.gl , you can see the libffi policy @ https://github.com/libffi/rlgl-policy/blob/master/XFAIL ), HOWEVER one big gap is around iOS testing.  I've never really participated in the iOS ecosystem.  If there was some way to run the testsuite on travis using a simulator, that would be wonderful.  Do you know if this is possible?

jeremyhu commented:
Sorry, time's been a bit more scarce for me recently.  I haven't had a chance to circle back to this.  I'm hoping to have some time this month to at least start forking off some of these changes into their own PRs and then look into the autoconf build system work.

jeremyhu commented:
#586 and #587 filed for some of the simpler changes made in Apple's libffi.  I'm hoping to break this out into other incremental PRs over the coming week or two.  Sorry this took longer to get back to than expected.

hjelmn commented:
@jeremyhu I can help finish the autotools changes if you like.

hjelmn commented:
For those who are interested. This is the code in libffi (src/aarch64/ffi.c) where the define is used:

```c
#ifdef FFI_TRAMPOLINE_WHOLE_DYLIB
  void **config = (void **)((uint8_t *)codeloc - 2*PAGE_MAX_SIZE);
#else
  void **config = (void **)((uint8_t *)codeloc - PAGE_MAX_SIZE);
#endif
  config[0] = closure;
  config[1] = start;
```

hjelmn commented:
Accidentally posted one extra comment (now deleted). Ran into the build system integration issue while trying to get libffi to work within ghc 8.10.3.

jeremyhu commented:
I split off some more preprocessing changes in #620, and I removed the whole-dylib trampoline changes such that this can focus on the arm64e changes.  I'll file a subsequent pull request with the `FFI_TRAMPOLINE_WHOLE_DYLIB` changes.

jeremyhu commented:
> @jeremyhu I can help finish the autotools changes if you like.

That would be greatly appreciated.  I've cleaned up the patches a bit.  This one and #620 are ready to merge.  The remaining piece is the trampoline changes.  I've split that out into  #621.  Please feel free to jump on that.  I hope I'll have time in the next few weeks to look at this again, but not sure.

jeremyhu commented:
Rebased to address merge conflicts with #624.

@atgreen Could we get this merged to avoid future merge conflicts?  Thanks.  The pieces that were not ready were split out into #621.
